codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -495,20 +518,22 @@ private void incrementPendingBytes(int numBytes) {
              * Note: this does not flush the {@link ChannelHandlerContext}.
              */
             int write(int allowedBytes) {
-                int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
                 try {
+                    pendingWriteQueue.remove();
```

Comment:
@buchgr - Can the we just remove the frame when it is actually "done" (fully written) before we notify the lifecycleManager?  Unless I'm mistaken otherwise it is an error and we may not want to ignore it.  Seems like we are separating the responsibility to notify the lifecycle manager and queue removal when they may be coupled.  We could push the queue down into the `FlowControlledBase` or have some interface to notify a listener of completion.  Where ever it is implemented the write completion could then be something like:

```java
public void completeWrite() {
  if (endOfStream) {
    pendingWriteQueue.remove();
    lifecycleManager.closeLocalSide(stream, promise);
  }
}
```